File: frontend/apps/web-antd/src/api/admin/tenant.ts
/**
 * 租户管理 API
 * 对接后端 /admin/tenants/* 接口
 */
import { requestClient } from '../request';

// ============================================================
// 类型定义
// ============================================================

/** 套餐类型 */
export type TenantPlan = 'free' | 'basic' | 'pro' | 'enterprise';

/** 租户列表查询参数 */
export interface TenantListParams {
  page?: number;
  page_size?: number;
  is_active?: boolean | null;
  plan?: TenantPlan | null;
}

/** 创建租户请求 */
export interface TenantCreateRequest {
  code: string;
  name: string;
  contact_name?: string | null;
  contact_phone?: string | null;
  contact_email?: string | null;
  plan?: TenantPlan;
  quota?: Record<string, any> | null;
  expires_at?: string | null;
  remark?: string | null;
}

/** 更新租户请求 */
export interface TenantUpdateRequest {
  name?: string | null;
  contact_name?: string | null;
  contact_phone?: string | null;
  contact_email?: string | null;
  plan?: TenantPlan | null;
  quota?: Record<string, any> | null;
  expires_at?: string | null;
  remark?: string | null;
}

/** 切换状态请求 */
export interface TenantStatusRequest {
  is_active: boolean;
}

/** 租户信息（后端原始格式 snake_case） */
export interface TenantInfoRaw {
  id: number;
  code: string;
  name: string;
  contact_name?: string;
  contact_phone?: string;
  contact_email?: string;
  plan: TenantPlan;
  quota?: Record<string, any>;
  is_active: boolean;
  expires_at?: string;
  remark?: string;
  created_at: string;
  updated_at?: string;
}

/** 租户信息（前端格式 camelCase） */
export interface TenantInfo {
  id: number;
  code: string;
  name: string;
  contactName?: string;
  contactPhone?: string;
  contactEmail?: string;
  plan: TenantPlan;
  quota?: Record<string, any>;
  isActive: boolean;
  expiresAt?: string;
  remark?: string;
  createdAt: string;
  updatedAt?: string;
}

/** 分页列表响应 */
export interface TenantListResponse {
  items: TenantInfo[];
  total: number;
  page: number;
  page_size: number;
}

// ============================================================
// 转换函数
// ============================================================

/** 将后端 snake_case 转换为前端 camelCase */
function transformTenantInfo(raw: TenantInfoRaw): TenantInfo {
  return {
    id: raw.id,
    code: raw.code,
    name: raw.name,
    contactName: raw.contact_name,
    contactPhone: raw.contact_phone,
    contactEmail: raw.contact_email,
    plan: raw.plan,
    quota: raw.quota,
    isActive: raw.is_active,
    expiresAt: raw.expires_at,
    remark: raw.remark,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

// ============================================================
// API 接口
// ============================================================

const API_PREFIX = '/admin/tenants';

/**
 * 获取租户列表
 * GET /admin/tenants
 */
export async function getTenantListApi(
  params?: TenantListParams,
): Promise<TenantListResponse> {
  const response = await requestClient.get<{
    items: TenantInfoRaw[];
    total: number;
    page: number;
    page_size: number;
  }>(API_PREFIX, { params });

  return {
    items: response.items.map(transformTenantInfo),
    total: response.total,
    page: response.page,
    page_size: response.page_size,
  };
}

/**
 * 获取租户详情
 * GET /admin/tenants/{tenant_id}
 */
export async function getTenantDetailApi(tenantId: number): Promise<TenantInfo> {
  const raw = await requestClient.get<TenantInfoRaw>(
    `${API_PREFIX}/${tenantId}`,
  );
  return transformTenantInfo(raw);
}

/**
 * 创建租户
 * POST /admin/tenants
 */
export async function createTenantApi(
  data: TenantCreateRequest,
): Promise<TenantInfo> {
  const raw = await requestClient.post<TenantInfoRaw>(API_PREFIX, data);
  return transformTenantInfo(raw);
}

/**
 * 更新租户
 * PUT /admin/tenants/{tenant_id}
 */
export async function updateTenantApi(
  tenantId: number,
  data: TenantUpdateRequest,
): Promise<TenantInfo> {
  const raw = await requestClient.put<TenantInfoRaw>(
    `${API_PREFIX}/${tenantId}`,
    data,
  );
  return transformTenantInfo(raw);
}

/**
 * 删除租户
 * DELETE /admin/tenants/{tenant_id}
 */
export async function deleteTenantApi(tenantId: number): Promise<void> {
  await requestClient.delete(`${API_PREFIX}/${tenantId}`);
}

/**
 * 切换租户状态
 * PUT /admin/tenants/{tenant_id}/status
 */
export async function toggleTenantStatusApi(
  tenantId: number,
  data: TenantStatusRequest,
): Promise<TenantInfo> {
  const raw = await requestClient.put<TenantInfoRaw>(
    `${API_PREFIX}/${tenantId}/status`,
    data,
  );
  return transformTenantInfo(raw);
}

File: backend/app/schemas/system/tenant.py
"""
租户相关 Schema

定义租户 API 的请求和响应数据结构
"""

from datetime import datetime
from typing import Any

from pydantic import Field

from app.core.base_schema import BaseSchema


class TenantResponse(BaseSchema):
    """租户信息响应"""
    
    id: int = Field(..., description="租户 ID")
    code: str = Field(..., description="租户编码")
    name: str = Field(..., description="租户名称")
    contact_name: str | None = Field(None, description="联系人姓名")
    contact_phone: str | None = Field(None, description="联系人电话")
    contact_email: str | None = Field(None, description="联系人邮箱")
    is_active: bool = Field(..., description="是否启用")
    plan: str = Field(..., description="套餐类型")
    quota: dict[str, Any] | None = Field(None, description="配额配置")
    expires_at: datetime | None = Field(None, description="到期时间")
    remark: str | None = Field(None, description="备注")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")


class TenantCreateRequest(BaseSchema):
    """创建租户请求"""
    
    code: str = Field(
        ..., 
        min_length=2, 
        max_length=50, 
        pattern=r"^[a-z][a-z0-9_-]*$",
        description="租户编码（唯一标识，小写字母开头，只能包含小写字母、数字、下划线、中划线）"
    )
    name: str = Field(..., min_length=1, max_length=100, description="租户名称")
    contact_name: str | None = Field(None, max_length=50, description="联系人姓名")
    contact_phone: str | None = Field(None, max_length=20, description="联系人电话")
    contact_email: str | None = Field(None, description="联系人邮箱")
    plan: str = Field("free", description="套餐类型")
    quota: dict[str, Any] | None = Field(None, description="配额配置")
    expires_at: datetime | None = Field(None, description="到期时间")
    remark: str | None = Field(None, max_length=500, description="备注")


class TenantUpdateRequest(BaseSchema):
    """更新租户请求"""
    
    name: str | None = Field(None, min_length=1, max_length=100, description="租户名称")
    contact_name: str | None = Field(None, max_length=50, description="联系人姓名")
    contact_phone: str | None = Field(None, max_length=20, description="联系人电话")
    contact_email: str | None = Field(None, description="联系人邮箱")
    plan: str | None = Field(None, description="套餐类型")
    quota: dict[str, Any] | None = Field(None, description="配额配置")
    expires_at: datetime | None = Field(None, description="到期时间")
    remark: str | None = Field(None, max_length=500, description="备注")


class TenantStatusRequest(BaseSchema):
    """租户状态切换请求"""
    
    is_active: bool = Field(..., description="是否启用")


class TenantImpersonateRequest(BaseSchema):
    """一键登录租户后台请求"""
    
    role_id: int | None = Field(None, description="目标角色 ID（可选）")


class TenantImpersonateResponse(BaseSchema):
    """一键登录租户后台响应"""
    
    impersonate_token: str = Field(..., description="一键登录 Token（60秒有效，一次性）")
    tenant_code: str = Field(..., description="租户编码")
    tenant_name: str = Field(..., description="租户名称")
    expires_in: int = Field(60, description="Token 有效期（秒）")


__all__ = [
    "TenantResponse",
    "TenantCreateRequest",
    "TenantUpdateRequest",
    "TenantStatusRequest",
    "TenantImpersonateRequest",
    "TenantImpersonateResponse",
]
